util/webhook/webhook.go

```diff
@@ -329,21 +329,42 @@ func (a *ArgoCDWebhookHandler) HandleEvent(payload interface{}) {
 	}
 }
 
-// getWebUrlRegex compiles a regex that will match any targetRevision referring to the same repo as the given webURL.
+// GetWebUrlRegex compiles a regex that will match any targetRevision referring to the same repo as the given webURL.
 // webURL is expected to be a URL from an SCM webhook payload pointing to the web page for the repo.
-func getWebUrlRegex(webURL string) (*regexp.Regexp, error) {
-	urlObj, err := url.Parse(webURL)
+func GetWebUrlRegex(webURL string) (*regexp.Regexp, error) {
+	// 1. Optional: protocol (`http`, `https`, or `ssh`) followed by `://`
+	// 2. Optional: username followed by `@`
+	// 3. Optional: `ssh` or `altssh` subdomain
+	// 4. Required: hostname parsed from `webURL`
+	// 5. Optional: `:` followed by port number
+	// 6. Required: `:` or `/`
+	// 7. Required: path parsed from `webURL`
+	// 8. Optional: `.git` extension
+	return getUrlRegex(webURL, `(?i)^((https?|ssh)://)?(%[1]s@)?((alt)?ssh\.)?%[2]s(:[0-9]+)?[:/]%[3]s(\.git)?$`)
+}
+
+// GetApiUrlRegex compiles a regex that will match any targetRevision referring to the same repo as the given apiURL.
+func GetApiUrlRegex(apiURL string) (*regexp.Regexp, error) {
+	// 1. Optional: protocol (`http` or `https`) followed by `://`
+	// 2. Optional: username followed by `@`
+	// 3. Required: hostname parsed from `webURL`
+	// 4. Optional: `:` followed by port number
+	// 5. Optional: `/`
+	return getUrlRegex(apiURL, `(?i)^(https?://)?(%[1]s@)?%[2]s(:[0-9]+)?/?$`)
```

Comment:
These regexes seem different from the previous ones.

Reply:
Correct, there are subtle differences.

The previous `Application` webhook web URL regex was:
```
(http://|https://|%s@|ssh://(%s@)?)%s(:[0-9]+|)[:/]%s(\.git)?
```

Which broke down to:
1. `http://`, `https://`, `username@`, `ssh://`, or `ssh://username@` (where `username` is the RFC 3986 pattern defined elsewhere in the file)
2. hostname from original URL, passed through `regexp.QuoteMeta()`
3. `:port` or nothing
4. colon or slash
5. path from original URL, passed through `regexp.QuoteMeta()` (which might be empty)
6. optional `.git` extension

The previous `ApplicationSet` webhook web URL regex was:
```
(http://|https://|\w+@|ssh://(\w+@)?)%s(:[0-9]+|)[:/]%s(\.git)?
```

Which differed from the `Application` one in that:
* Any sequence of one or more word characters was allowed for username.
* The hostname and path values were not passed through `regexp.QuoteMeta()`.

In the new common web URL regex used by both webhooks:
* Protocol and username are separated and are both optional. (This fixes #15700 to allow usernames with `https?` URLs.)
* An optional `altssh` or `ssh` subdomain is allowed. (This fixes #19314 to allow for SSH-over-443 alternatives.)
* Hostname and path are both passed through `regexp.QuoteMeta()`. (This doesn’t fix any reported issue that I’m aware of, but prevents edge cases of hostname/path values containing anything that would have otherwise been interpreted as regex syntax.)

The previous `ApplicationSet` webhook API URL regex was:
```
(http://|https://|\w+@|ssh://(\w+@)?)%s(:[0-9]+|)[:/]
```

The new API URL regex has the same benefits as listed above for the new web URL regex (except for the `altssh`/`ssh` subdomain allowance since these URLs will never have that).